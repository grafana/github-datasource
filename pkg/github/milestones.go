package github

import (
	"context"
	"time"

	"github.com/grafana/github-datasource/pkg/models"
	"github.com/grafana/grafana-plugin-sdk-go/data"
	"github.com/pkg/errors"
	"github.com/shurcooL/githubv4"
)

// QueryListMilestones lists all milestones in a repository
// {
//   repository(name: "grafana", owner: "grafana") {
//     milestones(first: 100) {
//       nodes {
//         color
//         description
//         name
//       }
//     }
//   }
// }
type QueryListMilestones struct {
	Repository struct {
		Milestones struct {
			Nodes    Milestones
			PageInfo PageInfo
		} `graphql:"milestones(first: 100, after: $cursor, query: $query)"`
	} `graphql:"repository(name: $name, owner: $owner)"`
}

// Milestone is a GitHub Milestone
type Milestone struct {
	Closed  bool
	Creator struct {
		User User `graphql:"... on User"`
	}
	DueOn     githubv4.DateTime
	ClosedAt  githubv4.DateTime
	CreatedAt githubv4.DateTime
	State     githubv4.MilestoneState
	Title     string
}

// Milestones is a list of GitHub milestones
type Milestones []Milestone

// Frames converts the list of GitHub Milestones to a Grafana data frame
func (m Milestones) Frames() data.Frames {
	frame := data.NewFrame(
		"milestones",
		data.NewField("title", nil, []string{}),
		data.NewField("author", nil, []string{}),
		data.NewField("closed", nil, []bool{}),
		data.NewField("state", nil, []string{}),
		data.NewField("created_at", nil, []time.Time{}),
		data.NewField("closed_at", nil, []*time.Time{}),
		data.NewField("due_at", nil, []*time.Time{}),
	)

	for _, v := range m {
		var (
			closedAt *time.Time
			dueAt    *time.Time
		)
		if !v.ClosedAt.Time.IsZero() {
			closedAt = &v.ClosedAt.Time
		}

		if !v.DueOn.Time.IsZero() {
			dueAt = &v.DueOn.Time
		}

		frame.AppendRow(
			v.Title,
			v.Creator.User.Login,
			v.Closed,
			string(v.State),
			v.CreatedAt.Time,
			closedAt,
			dueAt,
		)
	}

	return data.Frames{frame}
}

// GetAllMilestones lists milestones in a repository
func GetAllMilestones(ctx context.Context, client Client, opts models.ListMilestonesOptions) (Milestones, error) {
	var (
		variables = map[string]interface{}{
			"cursor": (*githubv4.String)(nil),
			"query":  githubv4.String(opts.Query),
			"owner":  githubv4.String(opts.Owner),
			"name":   githubv4.String(opts.Repository),
		}

		milestones = Milestones{}
	)

	for {
		q := &QueryListMilestones{}
		if err := client.Query(ctx, q, variables); err != nil {
			return nil, errors.WithStack(err)
		}

		milestones = append(milestones, q.Repository.Milestones.Nodes...)

		if !q.Repository.Milestones.PageInfo.HasNextPage {
			break
		}
		variables["cursor"] = q.Repository.Milestones.PageInfo.EndCursor
	}

	return milestones, nil
}
