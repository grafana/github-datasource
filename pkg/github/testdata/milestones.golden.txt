ðŸŒŸ This was machine generated.  Do not edit. ðŸŒŸ

Frame[0] 
Name: milestones
Dimensions: 7 Fields by 3 Rows
+------------------+----------------+--------------+----------------+---------------------------------+---------------------------------+---------------------------------+
| Name: title      | Name: author   | Name: closed | Name: state    | Name: created_at                | Name: closed_at                 | Name: due_at                    |
| Labels:          | Labels:        | Labels:      | Labels:        | Labels:                         | Labels:                         | Labels:                         |
| Type: []string   | Type: []string | Type: []bool | Type: []string | Type: []time.Time               | Type: []*time.Time              | Type: []*time.Time              |
+------------------+----------------+--------------+----------------+---------------------------------+---------------------------------+---------------------------------+
| first milestone  | testUser       | false        | OPEN           | 2020-08-25 16:21:56 +0000 +0000 | null                            | 2020-08-29 20:21:56 +0000 +0000 |
| seoncd milestone | testUser2      | true         | CLOSED         | 2020-08-25 16:21:56 +0000 +0000 | 2020-08-26 02:21:56 +0000 +0000 | 2020-08-29 20:21:56 +0000 +0000 |
| third milestone  | testUser2      | false        | OPEN           | 2020-08-25 16:21:56 +0000 +0000 | null                            | 2020-08-30 16:21:56 +0000 +0000 |
+------------------+----------------+--------------+----------------+---------------------------------+---------------------------------+---------------------------------+


====== TEST DATA RESPONSE (arrow base64) ======
FRAME=QVJST1cxAAD/////aAMAABAAAAAAAAoADgAMAAsABAAKAAAAFAAAAAAAAAEDAAoADAAAAAgABAAKAAAACAAAAFgAAAACAAAAKAAAAAQAAAAc/f//CAAAAAwAAAAAAAAAAAAAAAUAAAByZWZJZAAAADz9//8IAAAAFAAAAAoAAABtaWxlc3RvbmVzAAAEAAAAbmFtZQAAAAAHAAAAeAIAAAgCAACsAQAAUAEAAOAAAAB4AAAABAAAAKL///8UAAAAPAAAADwAAAAAAAoBPAAAAAEAAAAEAAAApP3//wgAAAAQAAAABgAAAGR1ZV9hdAAABAAAAG5hbWUAAAAAAAAAACb///8AAAMABgAAAGR1ZV9hdAAAAAASABgAFAATABIADAAAAAgABAASAAAAFAAAAEAAAABAAAAAAAAKAUAAAAABAAAABAAAABT+//8IAAAAFAAAAAkAAABjbG9zZWRfYXQAAAAEAAAAbmFtZQAAAAAAAAAAmv///wAAAwAJAAAAY2xvc2VkX2F0AAAAiv7//xQAAABAAAAASAAAAAAAAApIAAAAAQAAAAQAAAB4/v//CAAAABQAAAAKAAAAY3JlYXRlZF9hdAAABAAAAG5hbWUAAAAAAAAAAAAABgAIAAYABgAAAAAAAwAKAAAAY3JlYXRlZF9hdAAA9v7//xQAAAA8AAAAPAAAAAAAAAU4AAAAAQAAAAQAAADk/v//CAAAABAAAAAFAAAAc3RhdGUAAAAEAAAAbmFtZQAAAAAAAAAA3P7//wUAAABzdGF0ZQAAAE7///8UAAAAPAAAADwAAAAAAAAGOAAAAAEAAAAEAAAAPP///wgAAAAQAAAABgAAAGNsb3NlZAAABAAAAG5hbWUAAAAAAAAAADT///8GAAAAY2xvc2VkAACm////FAAAADwAAAA8AAAAAAAABTgAAAABAAAABAAAAJT///8IAAAAEAAAAAYAAABhdXRob3IAAAQAAABuYW1lAAAAAAAAAACM////BgAAAGF1dGhvcgAAAAASABgAFAAAABMADAAAAAgABAASAAAAFAAAAEQAAABIAAAAAAAABUQAAAABAAAADAAAAAgADAAIAAQACAAAAAgAAAAQAAAABQAAAHRpdGxlAAAABAAAAG5hbWUAAAAAAAAAAAQABAAEAAAABQAAAHRpdGxlAAAA/////9gBAAAUAAAAAAAAAAwAFgAUABMADAAEAAwAAADoAAAAAAAAABQAAAAAAAADAwAKABgADAAIAAQACgAAABQAAAAoAQAAAwAAAAAAAAAAAAAAEQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAEAAAAAAAAAAwAAAAAAAAAEAAAAAAAAAAAAAAAAAAAABAAAAAAAAAABAAAAAAAAAAUAAAAAAAAAAgAAAAAAAAAHAAAAAAAAAAAAAAAAAAAABwAAAAAAAAAAgAAAAAAAAAeAAAAAAAAAAAAAAAAAAAAHgAAAAAAAAAEAAAAAAAAACIAAAAAAAAABAAAAAAAAAAmAAAAAAAAAAAAAAAAAAAAJgAAAAAAAAAGAAAAAAAAACwAAAAAAAAAAgAAAAAAAAAuAAAAAAAAAAYAAAAAAAAANAAAAAAAAAAAAAAAAAAAADQAAAAAAAAABgAAAAAAAAAAAAAAAcAAAADAAAAAAAAAAAAAAAAAAAAAwAAAAAAAAAAAAAAAAAAAAMAAAAAAAAAAAAAAAAAAAADAAAAAAAAAAAAAAAAAAAAAwAAAAAAAAAAAAAAAAAAAAMAAAAAAAAAAgAAAAAAAAADAAAAAAAAAAAAAAAAAAAAAAAAAA8AAAAfAAAALgAAAGZpcnN0IG1pbGVzdG9uZXNlb25jZCBtaWxlc3RvbmV0aGlyZCBtaWxlc3RvbmUAAAAAAAAIAAAAEQAAABoAAAB0ZXN0VXNlcnRlc3RVc2VyMnRlc3RVc2VyMgAAAAAAAAIAAAAAAAAAAAAAAAQAAAAKAAAADgAAAE9QRU5DTE9TRURPUEVOAAAAaO2yVY8uFgBo7bJVjy4WAGjtslWPLhYCAAAAAAAAAAAAAAAAAAAAAKgjmhOwLhYAAAAAAAAAAADoC7vA1i8WAOgLu8DWLxYAaHiJPBgwFhAAAAAMABQAEgAMAAgABAAMAAAAEAAAACwAAAA8AAAAAAADAAEAAAB4AwAAAAAAAOABAAAAAAAA6AAAAAAAAAAAAAAAAAAAAAAAAAAAAAoADAAAAAgABAAKAAAACAAAAFgAAAACAAAAKAAAAAQAAAAc/f//CAAAAAwAAAAAAAAAAAAAAAUAAAByZWZJZAAAADz9//8IAAAAFAAAAAoAAABtaWxlc3RvbmVzAAAEAAAAbmFtZQAAAAAHAAAAeAIAAAgCAACsAQAAUAEAAOAAAAB4AAAABAAAAKL///8UAAAAPAAAADwAAAAAAAoBPAAAAAEAAAAEAAAApP3//wgAAAAQAAAABgAAAGR1ZV9hdAAABAAAAG5hbWUAAAAAAAAAACb///8AAAMABgAAAGR1ZV9hdAAAAAASABgAFAATABIADAAAAAgABAASAAAAFAAAAEAAAABAAAAAAAAKAUAAAAABAAAABAAAABT+//8IAAAAFAAAAAkAAABjbG9zZWRfYXQAAAAEAAAAbmFtZQAAAAAAAAAAmv///wAAAwAJAAAAY2xvc2VkX2F0AAAAiv7//xQAAABAAAAASAAAAAAAAApIAAAAAQAAAAQAAAB4/v//CAAAABQAAAAKAAAAY3JlYXRlZF9hdAAABAAAAG5hbWUAAAAAAAAAAAAABgAIAAYABgAAAAAAAwAKAAAAY3JlYXRlZF9hdAAA9v7//xQAAAA8AAAAPAAAAAAAAAU4AAAAAQAAAAQAAADk/v//CAAAABAAAAAFAAAAc3RhdGUAAAAEAAAAbmFtZQAAAAAAAAAA3P7//wUAAABzdGF0ZQAAAE7///8UAAAAPAAAADwAAAAAAAAGOAAAAAEAAAAEAAAAPP///wgAAAAQAAAABgAAAGNsb3NlZAAABAAAAG5hbWUAAAAAAAAAADT///8GAAAAY2xvc2VkAACm////FAAAADwAAAA8AAAAAAAABTgAAAABAAAABAAAAJT///8IAAAAEAAAAAYAAABhdXRob3IAAAQAAABuYW1lAAAAAAAAAACM////BgAAAGF1dGhvcgAAAAASABgAFAAAABMADAAAAAgABAASAAAAFAAAAEQAAABIAAAAAAAABUQAAAABAAAADAAAAAgADAAIAAQACAAAAAgAAAAQAAAABQAAAHRpdGxlAAAABAAAAG5hbWUAAAAAAAAAAAQABAAEAAAABQAAAHRpdGxlAAAAmAMAAEFSUk9XMQ==
